core/src/main/scala/ai/forestflow/utils/SourceStorageProtocols.scala

```diff
@@ -101,57 +101,91 @@ object SourceStorageProtocols extends StrictLogging {
-          case Success(value) => value
-          case _              => System.getenv(s"${credentialsRootKey}_$S3_ACCESS_KEY_ID_POSTFIX")
+        def getProcessedKey(key: String): String = {
+          key.replaceAll("""[^A-Za-z0-9]""", "_")
+            .replaceAll("""_+""", "_")
         }
 
-        val secretAccessKey = Try(s3Configs.getString(s"$credentialsRootKey.$S3_SECRET_ACCESS_KEY_POSTFIX")) match {
-          case Success(value) => value
-          case _              => System.getenv(s"${credentialsRootKey}_$S3_SECRET_ACCESS_KEY_POSTFIX")
-        }
+        def getS3Credentials(credentialsRootKey: String): Option[(String, String)] = {
 
-        require(accessKeyId != null && !accessKeyId.isEmpty, "S3 storage's 'Access Key' cannot be null or empty. Refer to the docs on how to configure this information" )
-        require(secretAccessKey != null && !secretAccessKey.isEmpty, "S3 storage's 'Access Key Secret' cannot be null or empty. Refer to the docs on how to configure this information" )
+          def getCredential(postFix: String): String = {
+            Try(s3Configs.getString(s"$credentialsRootKey.$postFix")) match {
+              case Success(value) => value
+              case _ => System.getenv(s"${credentialsRootKey}_$postFix") // System.getenv call returns null if the environment variable is not found
+            }
+          }
 
-        AwsBasicCredentials.create(accessKeyId,secretAccessKey)
-      }
+          val accessKeyId = getCredential(S3_ACCESS_KEY_ID_POSTFIX)
+          val secretAccessKey = getCredential(S3_SECRET_ACCESS_KEY_POSTFIX)
 
+          if (accessKeyId != null && secretAccessKey != null) {
+            Some((accessKeyId, secretAccessKey))
+          }
+          else {
+            None
+          }
+        }
+
+        //the format for the rootKey needs to be strictly followed - refer to the docs for more information
+        //the getProcessedKey, is a pre-processing step that applies the pre-defined format
+        val authorityProcessed = getProcessedKey(uri.getAuthority)
+        val pathProcessed = getProcessedKey(uri.getPath)
+        val artifactNameProcessed = getProcessedKey(artifactName)
+
+        // The keyLevels list represents the hierarchy to be followed for fall-back mechanism from left to right
+        // pathProcessed is empty for non-path style bucket access
+
+        val keyLevels = List(
+          s"$authorityProcessed${pathProcessed}_$artifactNameProcessed",   //url_port_bucket_key,
+          s"$authorityProcessed$pathProcessed",                            //url_port_bucket
+          s"$authorityProcessed"                                           //url_port
+        )
+
+        // the combination of map and find acts like collectFirst
+        val (accessKeyId, secretAccessKey) = keyLevels
+          .map(getS3Credentials)
+          .find(_.isDefined)
+          .flatten
+          .getOrElse(("", ""))
```

Comment:
This looks like a great candidate for Scala's collectFirst.
Here's an example from our own code
`ProtocolPatterns.collectFirst { case (protocol, reg) if reg.findFirstIn(path).isDefined => protocol }`